### ISSUE TITLE
[CHORE] Run hash generation on each PR?

### ISSUE THREAD

**Issue by Felienne** [hedyorg/hedy#4546]
We now periodically (daily) run a GitHub action to generate hashes for all snippets (See #3757 and #3873). I am not sure the why is documented somewhere so let me do it here briefly:

The repo contains a lot of code snippets, in /adventures and in /parsons for example, and we want to make sure those snippets don't have errors in them. Therefore we test them against the grammar files. But running all snippets tests takes a very long time (1.5 hours!) and that is annoying, esp for PRs that have nothing to do with the grammars.

So, we run the tests, and if we have seen a code snippet with this version of the Hedy language, we store the hash, so we know to skip it later on. That shrinks the time for tests from 90 minutes to about 10. Every night we regenerate all hashes so we remain up to date.
 
However, there are now 2 issues:
* The commits clog the history a bit
* (more important) if we merge a grammar PR, for the whole day the tests are slow until the evening (this happened yesterday for example)

So, we could also run the generation on each PR, but only do it when a grammar change was made (we used to have code for this, see https://github.com/hedyorg/hedy/pull/3895/files where it was removed). Doing it on each PR irrespective of changes to files would make each PR really slow. 

I once started work on this: https://github.com/hedyorg/hedy/pull/4078 but then I gave up because I was not sure enough about the Actions semantics, but it would add value.

**Comment by Felienne** [comment]
Potentially something that might interest @jtwaleson?

**Comment by jtwaleson** [comment]
Yes I like this a lot. I believe it makes sense to use github workflow caches which are designed for speeding up CI/CD pipelines. See https://docs.github.com/en/actions/using-workflows/caching-dependencies-to-speed-up-workflows

Compared to the current situation here are the cons:
- the hashes will not be present in the repo anymore (i.e. excluded with .gitignore)
- locally the tests will be slow as they do not re-use the cache
- for some branches it might not be possible to re-use the cache from the main branch, so the first time a pipeline runs on that branch it will be slow too (I have to investigate this in more detail, according to the docs it should be fine for most cases).

The pros are (in non prioritized order):
- remove big & frequently changing files from the codebase, reducing the time it takes to download the repo
- remove periodic commits that clutter the history
- faster CI/CD for most cases
- it's lazy, meaning that if there's no activity the cache won't be re-generated

Please let me know if you like this approach and I will give it a try.

**Comment by Felienne** [comment]
Hi @jtwaleson!

Thanks for looking into this in more depth! My plan was (as always, haha) a gradual one :) 

Can we start by changing the action into generating the hashes as we have them now, but on each PR (potentially with a filter for changed files)? That solves some of the problems already (just one extra commit in the history, for example). Plus it has the benefit that I think it can be done in a few lines in a GitHub action (and I think I was close in https://github.com/hedyorg/hedy/pull/4078)

From there, we can explore the cache. I am ok with most of the cons, but the fact that I will not be able to use the cache locally will be a time sink for me, so that must be avoided (since I most often actually need these tests). Is there a way to store the file elsewhere online and then grap it locally when needed? But I propose we dive into that avenue when we have first made the "on the PR" change.

**Comment by Felienne** [comment]
Now seeing that you also replied to #4541, thanks! Can we give #4541 priority, since that now leads to an issue for users in /slides and for content authors, and the happiness of both of those groups goes before devs :)

**Comment by jtwaleson** [comment]
Yes, will do.

**Comment by jtwaleson** [comment]
I looked into the code for hash generation etc and I understand how it works now. I agree and think it's best to do this on any PR and on the main branch with a filter on any file that impacts the hedy program (i.e. the files that impacts the hash, which we run in Tester.py).


Also some good news: with a couple tweaks I was able to shave off time for the regular 10-11 minute unit test run, which is now 4 minute 26 seconds: https://github.com/jtwaleson/hedy/actions/runs/6322684021/job/17168751032 . There's still some cleanup to do and also some room for further speed up. I can open a separate PR for that if desired.
